046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,63 @@
+<!-- This template is provided as an example with sections you may wish to comment on with respect to your proposal. Add or remove sections as required to best articulate the proposal. -->
+
+# Analyse broker metrics before rolling kafka
+
+This proposal describes a solution for KafkaRoller's shortcoming with brokers in log recovery. 
+
+## Current situation
+
+The current logic in Kafka Roller:
+- Take a list of kafka pods and reorder them based on the readiness. 
+- Roll unready pods first without considering why it could be unready.
+- Before restarting a kafka pod, consider if it is a controller or it has an impact on the availability.
+- If a kafka pod does not become ready within the operational timeout (300000ms by default) after restarting, then force restart it without considering why it's taking a long time.
+
+A Kafka broker can take a long time to become ready while performing log recovery. In this case, KafkaRoller ends up force restarting the broker and this can continue indefinetly, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
+
+KafkaAgent currently collects `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is `3`(`RUNNING`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation 
+
+To address the issue described above, the KafkaRoller needs to be aware of the current broker state, so that if a broker is performing log recovery, then it waits longer instead of force restarting it.
+
+## Proposal
+
+The proposal is a REST API endpoint on the KafkaAgent which exposes the BrokerState metric that it already uses. And modify KafkaRoller so that it queries the new endpoint to find out what state a broker is in. 
+
+1. Add a web server to KafkaAgent that exposes an endpoint which returns a value of metric
+    - Configure Jetty server to only accept SSL connections on a port 8443.
+    - Expose an endpoint `/broker-state`.
+    - Add a servlet to handle a `GET` request and return an integer value of `BrokerState` metric.
+    - It will reuse broker's own certificate, which will already be in the container.
```

Comment:
What about security beyond encryption? E.g. TLS client authentication? I wonder what value encryption adds without authentication.

Reply:
Are you suggesting that we utilise mTLS, that way we do both server and client authentications? Is mTLS enabled by default or configurable in Strimzi?

Reply:
All internal communication is secured with mTLS out of the box without any possibility to disable (at least right now).